File: internal/domain/window.go
// Package domain holds the pure logic of the dashboard: how a time window is
// chosen, how CloudWatch Logs Insights queries are built, and how AWS responses
// are shaped for the frontend. Nothing here talks to AWS or to HTTP, so all of
// it is directly testable.
package domain

import (
	"fmt"
	"strings"
	"time"
)

// Range is how far back a view looks. Four hours is a hard product limit.
type Range time.Duration

// Period is the width of one bucket on the time axis.
type Period time.Duration

const (
	Range15m Range = Range(15 * time.Minute)
	Range30m Range = Range(30 * time.Minute)
	Range1h  Range = Range(time.Hour)
	Range2h  Range = Range(2 * time.Hour)
	Range4h  Range = Range(4 * time.Hour)

	Period1m  Period = Period(time.Minute)
	Period5m  Period = Period(5 * time.Minute)
	Period10m Period = Period(10 * time.Minute)
	Period1h  Period = Period(time.Hour)
)

// MaxRange is the ceiling the product imposes on every log and metric query.
// Bounding the window is what keeps a Logs Insights scan predictable, so this
// is enforced server-side and not merely offered as the largest UI option.
const MaxRange = 4 * time.Hour

// A window must hold enough buckets to read as a series, and few enough that
// the browser is not asked to plot noise. Both bounds are inclusive.
const (
	minBuckets = 4
	maxBuckets = 250
)

var (
	allRanges  = []Range{Range15m, Range30m, Range1h, Range2h, Range4h}
	allPeriods = []Period{Period1m, Period5m, Period10m, Period1h}

	// defaultPeriods favours resolution while keeping bucket counts modest.
	defaultPeriods = map[Range]Period{
		Range15m: Period1m,
		Range30m: Period1m,
		Range1h:  Period1m,
		Range2h:  Period1m,
		Range4h:  Period5m,
	}
)

func (r Range) Duration() time.Duration  { return time.Duration(r) }
func (p Period) Duration() time.Duration { return time.Duration(p) }

// String renders the compact form used on the wire and in the URL: "15m", "4h".
func (r Range) String() string  { return compactDuration(time.Duration(r)) }
func (p Period) String() string { return compactDuration(time.Duration(p)) }

// Seconds is the unit CloudWatch's Period parameter expects.
func (p Period) Seconds() int32 { return int32(time.Duration(p) / time.Second) }

func compactDuration(d time.Duration) string {
	switch {
	case d%time.Hour == 0:
		return fmt.Sprintf("%dh", int(d/time.Hour))
	case d%time.Minute == 0:
		return fmt.Sprintf("%dm", int(d/time.Minute))
	default:
		return fmt.Sprintf("%ds", int(d/time.Second))
	}
}

// Ranges lists every selectable range, shortest first.
func Ranges() []Range { return append([]Range(nil), allRanges...) }

// ParseRange accepts the compact form and rejects anything outside the fixed
// set. Free-form durations are deliberately not supported: an arbitrary range
// would let a caller slip past MaxRange by a rounding error, and it would make
// the range/period compatibility table unbounded.
func ParseRange(s string) (Range, error) {
	for _, r := range allRanges {
		if r.String() == s {
			return r, nil
		}
	}
	d, err := time.ParseDuration(s)
	if err != nil {
		return 0, fmt.Errorf("unknown range %q", s)
	}
	if d > MaxRange {
		return 0, fmt.Errorf("range %s exceeds the %s maximum", s, compactDuration(MaxRange))
	}
	return 0, fmt.Errorf("unsupported range %q; allowed: %s", s, joinRanges(allRanges))
}

// ParsePeriod accepts the compact form of a bucket width.
func ParsePeriod(s string) (Period, error) {
	for _, p := range allPeriods {
		if p.String() == s {
			return p, nil
		}
	}
	return 0, fmt.Errorf("unsupported period %q; allowed: %s", s, joinPeriods(allPeriods))
}

// PeriodsFor lists the bucket widths that produce a sensible number of buckets
// for r. Callers use it to build the UI selector, so the frontend can never
// offer a combination the server would reject.
func PeriodsFor(r Range) []Period {
	var out []Period
	for _, p := range allPeriods {
		if n := buckets(r, p); n >= minBuckets && n <= maxBuckets {
			out = append(out, p)
		}
	}
	return out
}

// DefaultPeriod is the bucket width used when a request omits one.
func DefaultPeriod(r Range) Period {
	if p, ok := defaultPeriods[r]; ok {
		return p
	}
	// Fall back to the coarsest period that still clears minBuckets.
	valid := PeriodsFor(r)
	if len(valid) == 0 {
		return Period1m
	}
	return valid[len(valid)-1]
}

func buckets(r Range, p Period) int {
	if p <= 0 {
		return 0
	}
	return int(time.Duration(r) / time.Duration(p))
}

// Window is the single time window that every panel in one response shares.
//
// The reference implementation anchored each panel to its own time.Now(), so a
// dashboard could show a metric chart and a log chart describing windows that
// were seconds apart and bucketed differently. Building the window once at the
// edge of a request and threading it through is what makes two panels — and the
// overview and detail views of the same panel — agree.
type Window struct {
	Start  time.Time
	End    time.Time
	Range  Range
	Period Period
}

// NewWindow builds the window for r/p as of now.
//
// End is floored to a period boundary and Start is derived from it, so every
// bucket in the window is a complete one. The reference implementation used a
// raw now for the query bound while bucketing with (ts/width)*width, which left
// the newest bucket partially filled and the oldest one truncated — the UI then
// worked around it by reading the second-to-last bucket in one place and the
// last one in another, which is precisely how the same number came to differ
// between two spots on the same screen.
func NewWindow(now time.Time, r Range, p Period) (Window, error) {
	if err := Validate(r, p); err != nil {
		return Window{}, err
	}
	end := now.UTC().Truncate(p.Duration())
	return Window{
		Start:  end.Add(-r.Duration()),
		End:    end,
		Range:  r,
		Period: p,
	}, nil
}

// Validate reports whether r and p may be combined.
func Validate(r Range, p Period) error {
	if r.Duration() <= 0 {
		return fmt.Errorf("range must be positive")
	}
	if r.Duration() > MaxRange {
		return fmt.Errorf("range %s exceeds the %s maximum", r, compactDuration(MaxRange))
	}
	if p.Duration() <= 0 {
		return fmt.Errorf("period must be positive")
	}
	n := buckets(r, p)
	if n < minBuckets {
		return fmt.Errorf("period %s over range %s yields %d buckets, fewer than the %d minimum", p, r, n, minBuckets)
	}
	if n > maxBuckets {
		return fmt.Errorf("period %s over range %s yields %d buckets, more than the %d maximum", p, r, n, maxBuckets)
	}
	if time.Duration(r)%time.Duration(p) != 0 {
		return fmt.Errorf("period %s does not divide range %s evenly", p, r)
	}
	return nil
}

// Buckets is the number of points on the time axis.
func (w Window) Buckets() int { return buckets(w.Range, w.Period) }

// Timestamps lists the left edge of every bucket as Unix seconds, ascending.
// This is the x axis the frontend plots against and the index every series in
// the response is aligned to.
func (w Window) Timestamps() []int64 {
	n := w.Buckets()
	out := make([]int64, n)
	step := int64(w.Period.Duration() / time.Second)
	start := w.Start.Unix()
	for i := range out {
		out[i] = start + int64(i)*step
	}
	return out
}

// Index returns the bucket t falls in, and whether it falls in the window at
// all. Points outside are dropped rather than clamped: clamping would silently
// pile out-of-window data onto the edge buckets and inflate them.
func (w Window) Index(t time.Time) (int, bool) {
	if t.Before(w.Start) || !t.Before(w.End) {
		return 0, false
	}
	return int(t.Sub(w.Start) / w.Period.Duration()), true
}

// Contains reports whether t falls inside the window.
func (w Window) Contains(t time.Time) bool {
	_, ok := w.Index(t)
	return ok
}

// Align snaps t down to its bucket's left edge.
func (w Window) Align(t time.Time) time.Time {
	return t.UTC().Truncate(w.Period.Duration())
}

// Equal reports whether two windows describe the same span at the same
// resolution. Handlers assert this across the panels they assemble.
func (w Window) Equal(o Window) bool {
	return w.Start.Equal(o.Start) && w.End.Equal(o.End) && w.Period == o.Period
}

func (w Window) String() string {
	return fmt.Sprintf("%s..%s/%s", w.Start.Format(time.RFC3339), w.End.Format(time.RFC3339), w.Period)
}

// Resolve turns raw query-string values into a window, filling in defaults.
// An empty range means the 1h default; an empty period means the default for
// the resolved range.
func Resolve(now time.Time, rangeStr, periodStr string) (Window, error) {
	r := Range1h
	if rangeStr != "" {
		parsed, err := ParseRange(rangeStr)
		if err != nil {
			return Window{}, err
		}
		r = parsed
	}
	p := DefaultPeriod(r)
	if periodStr != "" {
		parsed, err := ParsePeriod(periodStr)
		if err != nil {
			return Window{}, err
		}
		p = parsed
	}
	return NewWindow(now, r, p)
}

func joinRanges(rs []Range) string {
	s := make([]string, len(rs))
	for i, r := range rs {
		s[i] = r.String()
	}
	return strings.Join(s, ", ")
}

func joinPeriods(ps []Period) string {
	s := make([]string, len(ps))
	for i, p := range ps {
		s[i] = p.String()
	}
	return strings.Join(s, ", ")
}

File: server/src/domain/window.ts
// 시간 창을 고르는 규칙. internal/domain/window.go 의 이식이다.
//
// Go 는 time.Duration(나노초)으로 다뤘고 여기서는 밀리초로 다룬다. JS 의 Date
// 가 밀리초라 그쪽에 맞추는 편이 변환을 없앤다.

import { parseGoDuration } from './logfmt.ts';

const minute = 60_000;
const hour = 60 * minute;

/** 화면이 얼마나 뒤까지 보는지. 네 시간은 제품이 정한 단단한 상한이다. */
export type Range = number;
/** 시간축에서 한 칸의 너비. */
export type Period = number;

export const range15m: Range = 15 * minute;
export const range30m: Range = 30 * minute;
export const range1h: Range = hour;
export const range2h: Range = 2 * hour;
export const range4h: Range = 4 * hour;

export const period1m: Period = minute;
export const period5m: Period = 5 * minute;
export const period10m: Period = 10 * minute;
export const period1h: Period = hour;

/**
 * maxRange 는 모든 로그·메트릭 조회에 제품이 씌우는 천장이다. 창을 묶어 두는
 * 것이 Insights 스캔량을 예측 가능하게 만들므로, 이것은 UI 의 가장 큰 선택지로
 * 제시되는 데 그치지 않고 서버에서 강제된다.
 */
export const maxRange = 4 * hour;

// 창은 시계열로 읽힐 만큼 칸이 많아야 하고, 브라우저에 잡음을 그리게 하지 않을
// 만큼 적어야 한다. 양쪽 모두 포함이다.
const minBuckets = 4;
const maxBuckets = 250;

const allRanges: Range[] = [range15m, range30m, range1h, range2h, range4h];
const allPeriods: Period[] = [period1m, period5m, period10m, period1h];

/** 해상도를 살리되 칸 수를 얌전하게 두는 기본값. */
const defaultPeriods = new Map<Range, Period>([
	[range15m, period1m],
	[range30m, period1m],
	[range1h, period1m],
	[range2h, period1m],
	[range4h, period5m]
]);

/** 와이어와 URL 에서 쓰는 짧은 표기: "15m", "4h". */
export function compactDuration(ms: number): string {
	if (ms % hour === 0) return `${Math.trunc(ms / hour)}h`;
	if (ms % minute === 0) return `${Math.trunc(ms / minute)}m`;
	return `${Math.trunc(ms / 1000)}s`;
}

/** CloudWatch 의 Period 파라미터가 기대하는 단위. */
export function periodSeconds(p: Period): number {
	return Math.trunc(p / 1000);
}

/** 고를 수 있는 범위 전부. 짧은 것부터. */
export function ranges(): Range[] {
	return [...allRanges];
}

function joinDurations(list: number[]): string {
	return list.map(compactDuration).join(', ');
}

/**
 * parseRange 는 짧은 표기를 받고 정해진 집합 밖은 거절한다. 자유 형식 기간을
 * 일부러 지원하지 않는다 — 임의의 범위는 반올림 오차로 maxRange 를 빠져나갈
 * 여지를 주고, 범위·주기 호환표를 무한히 만든다.
 */
export function parseRange(s: string): Range {
	for (const r of allRanges) {
		if (compactDuration(r) === s) return r;
	}
	const ns = parseGoDuration(s);
	if (ns === null) throw new Error(`unknown range "${s}"`);
	if (ns / 1e6 > maxRange) {
		throw new Error(`range ${s} exceeds the ${compactDuration(maxRange)} maximum`);
	}
	throw new Error(`unsupported range "${s}"; allowed: ${joinDurations(allRanges)}`);
}

/** parsePeriod 는 칸 너비의 짧은 표기를 받는다. */
export function parsePeriod(s: string): Period {
	for (const p of allPeriods) {
		if (compactDuration(p) === s) return p;
	}
	throw new Error(`unsupported period "${s}"; allowed: ${joinDurations(allPeriods)}`);
}

function bucketCount(r: Range, p: Period): number {
	if (p <= 0) return 0;
	return Math.trunc(r / p);
}

/**
 * periodsFor 는 r 에 대해 말이 되는 칸 수를 내는 너비를 나열한다. UI 선택기를
 * 이것으로 만들므로, 프론트는 서버가 거절할 조합을 제시할 수 없다.
 */
export function periodsFor(r: Range): Period[] {
	return allPeriods.filter((p) => {
		const n = bucketCount(r, p);
		return n >= minBuckets && n <= maxBuckets;
	});
}

/** 요청이 주기를 빼먹었을 때 쓰는 칸 너비. */
export function defaultPeriod(r: Range): Period {
	const p = defaultPeriods.get(r);
	if (p !== undefined) return p;
	// minBuckets 를 넘기는 가장 성긴 주기로 물러난다.
	const valid = periodsFor(r);
	return valid.length === 0 ? period1m : (valid[valid.length - 1] as Period);
}

/**
 * Window 는 한 응답의 모든 패널이 공유하는 단 하나의 시간 창이다.
 *
 * 이전 구현은 패널마다 제 나름의 now 를 잡았다. 그래서 한 대시보드가 몇 초씩
 * 어긋나고 칸도 다르게 잘린 창을 설명하는 메트릭 차트와 로그 차트를 함께
 * 보여 줄 수 있었다. 요청의 가장자리에서 창을 한 번 만들어 아래로 꿰는 것이,
 * 두 패널이 — 그리고 같은 패널의 오버뷰와 상세가 — 일치하게 만드는 방법이다.
 *
 * start·end 는 epoch 밀리초다.
 */
export interface Window {
	start: number;
	end: number;
	range: Range;
	period: Period;
}

/**
 * newWindow 는 now 기준으로 r/p 의 창을 만든다.
 *
 * end 를 주기 경계로 내리고 start 를 거기서 끌어내므로, 창 안의 모든 칸이 완성된
 * 칸이다. 이전 구현은 조회 경계에 날것의 now 를 쓰면서 칸은 (ts/너비)*너비 로
 * 잘랐고, 그래서 가장 최근 칸은 반쯤 차고 가장 오래된 칸은 잘려 있었다. UI 는
 * 한 곳에서는 끝에서 두 번째 칸을, 다른 곳에서는 마지막 칸을 읽는 것으로 그것을
 * 우회했는데, 같은 숫자가 한 화면의 두 자리에서 달라진 경위가 바로 그것이다.
 */
export function newWindow(now: Date, r: Range, p: Period): Window {
	validateWindow(r, p);
	// Go 의 time.Truncate 는 서기 1년 기준으로 내림한다. epoch 는 그 기준에서
	// 날짜 단위로 떨어지고 여기 쓰는 주기는 모두 하루를 정확히 나누므로,
	// epoch 밀리초를 내림하는 것과 결과가 같다.
	const end = Math.floor(now.getTime() / p) * p;
	return { start: end - r, end, range: r, period: p };
}

/** validateWindow 는 r 과 p 를 함께 쓸 수 있는지 본다. */
export function validateWindow(r: Range, p: Period): void {
	if (r <= 0) throw new Error('range must be positive');
	if (r > maxRange) {
		throw new Error(`range ${compactDuration(r)} exceeds the ${compactDuration(maxRange)} maximum`);
	}
	if (p <= 0) throw new Error('period must be positive');

	const n = bucketCount(r, p);
	if (n < minBuckets) {
		throw new Error(
			`period ${compactDuration(p)} over range ${compactDuration(r)} yields ${n} buckets, fewer than the ${minBuckets} minimum`
		);
	}
	if (n > maxBuckets) {
		throw new Error(
			`period ${compactDuration(p)} over range ${compactDuration(r)} yields ${n} buckets, more than the ${maxBuckets} maximum`
		);
	}
	if (r % p !== 0) {
		throw new Error(
			`period ${compactDuration(p)} does not divide range ${compactDuration(r)} evenly`
		);
	}
}

/** 시간축의 점 개수. */
export function buckets(w: Window): number {
	return bucketCount(w.range, w.period);
}

/**
 * timestamps 는 모든 칸의 왼쪽 경계를 Unix 초로 오름차순 나열한다. 프론트가
 * 그리는 x 축이고, 응답의 모든 시계열이 맞춰지는 인덱스다.
 */
export function timestamps(w: Window): number[] {
	const n = buckets(w);
	const step = w.period / 1000;
	const start = Math.trunc(w.start / 1000);
	const out: number[] = new Array<number>(n);
	for (let i = 0; i < n; i++) out[i] = start + i * step;
	return out;
}

/**
 * indexOf 는 t 가 들어가는 칸과, 애초에 창 안에 들어가기는 하는지를 준다. 밖의
 * 점은 자르지 않고 버린다 — 자르면 창 밖 데이터가 가장자리 칸에 조용히 쌓여
 * 그 칸을 부풀린다.
 */
export function indexOf(w: Window, t: Date | number): number | null {
	const ms = typeof t === 'number' ? t : t.getTime();
	if (ms < w.start || ms >= w.end) return null;
	return Math.trunc((ms - w.start) / w.period);
}

/** contains 는 t 가 창 안에 드는지 답한다. */
export function contains(w: Window, t: Date | number): boolean {
	return indexOf(w, t) !== null;
}

/** align 은 t 를 그 칸의 왼쪽 경계로 내린다. */
export function align(w: Window, t: Date | number): number {
	const ms = typeof t === 'number' ? t : t.getTime();
	return Math.floor(ms / w.period) * w.period;
}

/**
 * windowsEqual 은 두 창이 같은 구간을 같은 해상도로 말하는지 본다. 핸들러가
 * 조립한 패널을 가로질러 이것을 확인한다.
 */
export function windowsEqual(a: Window, b: Window): boolean {
	return a.start === b.start && a.end === b.end && a.period === b.period;
}

export function windowString(w: Window): string {
	return `${new Date(w.start).toISOString()}..${new Date(w.end).toISOString()}/${compactDuration(w.period)}`;
}

/**
 * resolveWindow 는 쿼리 문자열의 날값을 창으로 바꾸고 기본값을 채운다. 범위가
 * 비어 있으면 1시간, 주기가 비어 있으면 그 범위의 기본 주기다.
 */
export function resolveWindow(now: Date, rangeStr: string, periodStr: string): Window {
	const r = rangeStr === '' ? range1h : parseRange(rangeStr);
	const p = periodStr === '' ? defaultPeriod(r) : parsePeriod(periodStr);
	return newWindow(now, r, p);
}
